### PR TITLE
Refactor Mana Calculation

### LINKED ISSUE #831: Update calculation methods for Consensus & Access Mana Vectors
Modify base mana vector implementation.
There should be a base mana vector interface, and two types of base mana vectors that implement it.
 1. `Consensus Base Mana Vector`: calculated only via Mana 1.
 2. `Access Base Mana Vector`: calculated via Mana 2

Then, modify mana plugin to use these.

### PR DESCRIPTION
**Pull request by lzpap**
Fixes #831 

- Introduce `BaseMana` and `BaseManaVector` interfaces.
- `AccessBaseMana` and `AccessBaseManaVector` implement `Mana2` calculation method.
- `ConsensusBaseMana` and `ConsensusBaseManaVector` implement `Mana1` calculation method.
- `WeightedBaseMana` and `WeightedBaseManaVector` implement the combination of `Mana1` and `Mana2` methods. `weight` parameter is used to fine-tune the weights of `Mana1` and `Mana2` components in the output. `WeightedBaseMana` is for research purposes and has a `targetManaType` that defines if it takes into account `Access` or `Consensus` `pledgeIDs`.
- Modify `PersistableBaseMana` to handle new interfaces.
- Mana plugin uses refactored vectors for `Access` and `Consensus` mana calculations.
- Fix mana references throughout the codebase.
- Change Websocket msg struct definition for mana related events.
- Adjust Grafana dashboard, prometheus and metrics.
- Add tests for mana package.
- Fix mana research dashboard

### CODE REVIEW

File: packages/mana/weightedbasevector.go [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/iotaledger/hive.go/identity"
+	"golang.org/x/xerrors"
+)
+
+// WeightedBaseManaVector represents a base mana vector
```
Comment (acha-bill) [inline packages/mana/weightedbasevector.go:13]:
```suggestion
// WeightedBaseManaVector represents a base mana vector.
```

File: packages/mana/weightedbasevector.go [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+// Size returns the size of this mana vector.
+func (w *WeightedBaseManaVector) Size() int {
+	w.RLock()
+	defer w.RUnlock()
+	return len(w.vector)
+}
+
+// SetWeight sets the weight for the while vector.
+func (w *WeightedBaseManaVector) SetWeight(weight float64) error {
+	if weight < 0.0 || weight > 1.0 {
```
Comment (acha-bill) [inline packages/mana/weightedbasevector.go:40]: Can these min and max weights be constants?

File: plugins/dashboard/manafeed.go [hunk 3]
```diff
@@ -149,7 +149,7 @@ func sendManaMapOnline() {
 		Type: MsgTypeManaMapOnline,
 		Data: accessPayload,
 	})
-	consensusManaList, err := manaPlugin.GetOnlineNodes(mana.AccessMana, mana.Mixed)
+	consensusManaList, err := manaPlugin.GetOnlineNodes(mana.AccessMana)
```
Comment (acha-bill) [inline plugins/dashboard/manafeed.go:152]: Should this not be `mana.ConsensusMana`?

**Review by acha-bill — CHANGES_REQUESTED** [review]
Good work!

File: packages/mana/accessbasevector_test.go [hunk 4]
```diff
@@ -74,30 +74,27 @@ func randNodeID() identity.ID {
...
-	assert.Equal(t, AccessMana, bmvAccess.vectorType)
-	assert.Equal(t, map[identity.ID]*BaseMana{}, bmvAccess.vector)
-
-	bmvCons := NewBaseManaVector(ConsensusMana)
-	assert.Equal(t, ConsensusMana, bmvCons.vectorType)
-	assert.Equal(t, map[identity.ID]*BaseMana{}, bmvCons.vector)
+func TestNewBaseManaVector_Access(t *testing.T) {
+	bmvAccess, err := NewBaseManaVector(AccessMana)
+	assert.NoError(t, err)
+	assert.Equal(t, AccessMana, bmvAccess.Type())
```
[the remark this answers is not in the document]
Comment (lzpap) [reply packages/mana/accessbasevector_test.go:80]: The `Type` testcase's purpose is to test the `Type()` method of an `AccessBaseManaVector`, while `TestNewBaseManaVector_Access`  tests if the correct map was created when calling for a new base mana vector.

File: packages/mana/weightedbasevector.go (hunk 2 above)
Comment (lzpap) [reply packages/mana/weightedbasevector.go:40]: Yes, they can.

File: plugins/dashboard/manafeed.go (hunk 3 above)
Comment (lzpap) [reply plugins/dashboard/manafeed.go:152]: Good catch!

File: packages/mana/accessbasevector_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (acha-bill) [reply packages/mana/accessbasevector_test.go:80]: ah ok

**Comment by acha-bill** [comment]
You still have one unchecked box
> fix mana research dashboard

Maybe you forgot to check it?

**Comment by lzpap** [comment]
> You still have one unchecked box
> 
> > fix mana research dashboard
> 
> Maybe you forgot to check it?

I'm working on it ;)

**Review by acha-bill — APPROVED** [review]

**Pull request merged**